Nü Onboarding: Mark III

jleandroperez opened the pull request:
### Details:
This PR implements the new Authentication Interface, along with a new class, in charge of wrapping Simperium and OnePassword interactions.

cc @bummytime 
cc @SylvesterWilmott This one definitely could use your feedback sir!!

Ref. #359 


---

### Scenario: UX
- [ ] Verify rotations are disabled in every screen.
- [ ] Verify the Input Text Views's outer border (in both SignIn and SignUp screens) is slightly different when they become the first responder.
- [ ] Verify the Keyboard always matches the kind of input text we're expecting.
- [ ] Verify the Reveal Password button effectively displays the (hidden) text. Upon release, the password should get hidden again.
- [ ] Verify that the Primary Action Button (SignIn / SignUp) is grayed out whenever the input is actually invalid.
- [ ] Verify the Authentication UI **is not dismissable** in iOS 13.
- [ ] Verify that both **SignUp / SignIn** display a spinner during activity.
- [ ] Verify the new UI reacts well in iPhone / iPad devices (including multitasking).
- [ ] Verify that **no matter what the System Theme is**, both Onboarding and Authentication UI's will be in light mode (with the statusBar visible, and navigation title in black, when applicable)

### Scenario: Login
1. Launch Simplenote
2. Press over the LogIn Button
3. Enter a set of valid credentials
4. Verify you're shortly after into the app

**Please: Repeat with 1Password fill-in**

### Scenario: SignUp
1. Launch Simplenote
2. Press over the SignUp Button
3. Enter a set of valid credentials
4. Verify you're shortly after into the app, and that a new user was effectively created

**Please: Repeat with 1Password fill-in**

### Scenario: 1Password Availability
1. Launch the app in the Simulator
2. Verify 1P is unavailable
3. Launch the app in a device with 1P installed
4. Verify the 1P button becomes visible, in both, SignIn and SignUp screens

### Scenario: Signup fallback to Login
1. Launch Simplenote and open the **Signup** Screen
2. Enter a set of valid credentials and press Signup
3. Verify that an Alert shows up, offering to Login Instead
4. Press over the **Sign In** button
5. Verify the Login UI shows up onScreen, and shortly after, you're into the app

### Scenario: Warnings
1. Launch Simplenote and open the **Login** Screen
2. Press over the Username field and press **Next**
3. Verify a warning shows up onscreen
4. Type a valid username
5. Verify that the second the username is valid, the warning goes away on its own
6. Repeat the same steps for the Password field

--- 

### Screenshot Time!

![Simulator Screen Shot - iPhone 8 - 2019-08-06 at 12 57 42](https://user-images.githubusercontent.com/1195260/62555732-d8be8a00-b849-11e9-9694-66153a336513.png)
![Simulator Screen Shot - iPhone 8 - 2019-08-06 at 12 57 45](https://user-images.githubusercontent.com/1195260/62555740-dbb97a80-b849-11e9-8a76-51c80a0985ed.png)
![Simulator Screen Shot - iPhone 8 - 2019-08-06 at 12 57 50](https://user-images.githubusercontent.com/1195260/62555742-deb46b00-b849-11e9-9a7e-65ead8cfe752.png)

bummytime commented:
@jleandroperez still reviewing, but 2 quick comments:

1) I noticed this validation issue while testing (ignore the keyboard weirdness, on device the keyboard appears to behave ok):

![validation](https://user-images.githubusercontent.com/154014/62646479-12f45e00-b914-11e9-8acb-3b82417189f9.gif)

2) On my iPad with iOS 13b5, I see this crasher when I tap the 1Password icon:

```
2019-08-07 13:09:09.273389-0500 Simplenote[1156:410602] *** Assertion failure in -[OnePasswordExtension activityViewControllerForItem:viewController:sender:typeIdentifier:], /Users/bummytime/Projects/simplenote-ios/Pods/1PasswordExtension/OnePasswordExtension.m:507
2019-08-07 13:09:09.274539-0500 Simplenote[1156:410602] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'sender must not be nil on iPad.'
*** First throw call stack:
(0x19d294df8 0x19cfbe0a4 0x19d1977f8 0x19d5cb6d4 0x104d66c90 0x104d61614 0x1044d33a4 0x104504738 0x104504a14 0x1a12651e0 0x1a0cb7c80 0x1a0cb7fe4 0x1a0cb7000 0x1a129ddec 0x1a129f12c 0x1a127bff4 0x1a12f2570 0x1a12f4acc 0x1a12eda40 0x19d212cd8 0x19d212c30 0x19d2123c8 0x19d20d514 0x19d20cdb4 0x1a751c328 0x1a1263e94 0x104478ab0 0x19d098424)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb) 
```

jleandroperez commented:
@bummytime Both issues addressed, should be good now. Thank you!!

bummytime commented:
@jleandroperez thank you! 1Password and validation issues are indeed fixed. 🙇 

Re: the keyboard funkiness I referenced above...it appears to be an actual issue primarily on iPads (both iOS 12.4 and iPadOS 13b5). FWIW I have seen the same flicker on iPhone occasionally too.

![ipad_ios13_keyboard_issue](https://user-images.githubusercontent.com/154014/62647917-484e7b00-b917-11e9-82f8-ce259f463c10.gif)

**To repro:**

0. Start with an external keyboard attached to an iPad
1. Launch app on device (or sim) and navigate to login/signup
2. Email field should be 1st responder...now tap the password field
3. Tap the email field again
4. Notice the keyboard flashes
5. Disconnect the external keyboard
6. Tap between the fields again
7. Notice the soft keyboard occasionally flashes

bummytime commented:
@jleandroperez SORRY!!!! One other thing I found in testing:

<img width="408" alt="Screen Shot 2019-08-07 at 1 51 34 PM" src="https://user-images.githubusercontent.com/154014/62649484-9add6680-b91a-11e9-9811-5028ea2a87d7.png">

(iOS 10 on iPhone 7 sim)

jleandroperez commented:
### Strange Keyboard Behavior:

For the record, the source of such issue was this console warning:

```
2019-08-08 12:38:54.464413-0300 Simplenote[69899:13071236] [AutoFill] Cannot show Automatic Strong Passwords for app bundleID: com.codality.NotationalFlow.Development due to error: iCloud Keychain is disabled
```

Fixed by enabling iCloud in the AppId. If it's still happening on your end, please, redownload the provisioning profiles!

### iOS 7 Navbar Overlap

Fixed via [this commit](https://github.com/Automattic/simplenote-ios/pull/385/commits/16767d7ac9acafe9e4900ec384fc8bf6f1c1e532)

---

**Ready for another look, thank you!!!**

jleandroperez commented:
Invoking @frosty or @aerych for another set of eyes. This one is a bit delicate!!!

Thanks everyone in advance!!

aerych commented on Simplenote/Classes/UIView+Animations.swift:
```diff
@@ -0,0 +1,20 @@
+//
+extension UIView {
+
+    /// Indicates if the receiver has the horizontal compact trait
```
Maybe update the comment?  It looks like it was copied from `UIView+Simplenote.swift`.

jleandroperez replied:
Nice catch, fixed, thank you!!

aerych commented on Simplenote/Classes/SPOnboardingViewController.swift:
```diff
@@ -99,15 +111,19 @@ private extension SPOnboardingViewController {
+
+    func presentAuthenticationInterface(mode: AuthenticationMode) {
+        guard let simperiumAuthenticator = authenticator else {
+            fatalError()
```
Maybe provide a description? Or log the issue before calling `fatalError`?

jleandroperez replied:
Added a description! Thank you!!

aerych commented:
<img width="728" alt="Screen Shot 2019-08-08 at 1 43 08 PM" src="https://user-images.githubusercontent.com/1435271/62728915-85cc0a80-b9e2-11e9-9a84-1ecb324d3e06.png">

If I rotate the iPad right after tapping login or sign up while the controllers are animating I can get the form to appear off-center.

aerych commented:
Findings from testing in the simulator: 

Scenario: UX
- Confirming that rotations are disabled in every screen on the iPhone, but do rotate on the iPad.
- Confirming TextView border is darker when its the first responder.
- Keyboard does match input type
- Reveal password button works but there is an issue with the magnifier.
- Primary action button is disabled until form is valid.
- Confirming a spinner appears in the primary action button.
- Confirming the UI works on both iOS and iPad. Tested multitasking with no issues.  Did find an issue with rotating while transitioning between views.

Scenario: Login 
- Works as described.

Scenario SignUp
- Works as described.

Scenario: 1Password Availability
- 1P does not appear in the simulator.

Scenario: Signup fallback to Login
- Works as described but left a comment about using Sign In vs Log In for the button text.

Scenario: Warnings
- Confirming that warnings appear when I try to move between fields by tapping the return button on the keyboard. I can still move between text fields by tapping on the fields directly and no warnings appear in this case.

jleandroperez commented:
✅  Fallback to LogIn text updated. Thanks for the suggestion!!
✅  I've updated the **Reveal Password** logic so that it's not a *reveal while pressed*, but instead, works now as a toggle. Cannot repro anymore the magnifier issues!

**Pending:**
I'm still working on the layout issues during push / rotation. Not sure where / how to patch that one!!

Thanks for the detailed review sir!!

frosty commented:
I haven't had chance to review this yet, but... in [this screenshot](https://user-images.githubusercontent.com/1195260/62555740-dbb97a80-b849-11e9-8a76-51c80a0985ed.png), are the error labels different colors?

jleandroperez commented:
@frosty error labels should be red! (Unless I'm missing something?). Do you mean... if each error label should have a different color?

jleandroperez commented:
@frosty You don't have human eyes. Not fooling anyone!!!

Turns out the Warning label(s) color was set in Interface Builder. I've:

1. Wired the **Simplenote Lipstick** color (actual name from Zeplin!)
2. And now the Input Text Fields get their borders red, whenever the payload is invalid

Thanks for spotting that!!

jleandroperez commented:
### Update:
Only pending is the rotation issue, which happens only in iOS 12's simulator. Leaving that one as is, _for now_, will attack that issue as a separate thing.

Thank you @aerych !!!!

jleandroperez commented:
@aerych @frosty @bummytime **thank you gentlemen**!!! merging this one.

Woooooo!!!
